agent/bls.mjs

```diff
@@ -0,0 +1,2976 @@
+            cc.debug( "Discovered BLS threshold is " ) +
+            cc.info( nThreshold ) + cc.debug( "." ) +
+            "\n" );
+        details.write( strLogPrefix +
+            cc.debug( "Discovered number of BLS participants is " ) +
+            cc.info( nParticipants ) + cc.debug( "." ) +
+            "\n" );
+        details.write( strLogPrefix +
+            cc.debug( "hash value to sign is " ) + cc.info( strMessageHash ) +
+            "\n" );
+        let joAccount = imaState.chainProperties.sc.joAccount;
+        if( ! joAccount.strURL ) {
+            joAccount = imaState.chainProperties.mn.joAccount;
+            if( ! joAccount.strSgxURL )
+                throw new Error( "SGX URL is unknown, cannot sign U256" );
+            if( ! joAccount.strBlsKeyName )
+                throw new Error( "BLS keys name is unknown, cannot sign U256" );
+        }
+        let rpcCallOpts = null;
+        if( "strPathSslKey" in joAccount &&
+            typeof joAccount.strPathSslKey == "string" &&
+            joAccount.strPathSslKey.length > 0 &&
+            "strPathSslCert" in joAccount &&
+            typeof joAccount.strPathSslCert == "string" &&
+            joAccount.strPathSslCert.length > 0
+        ) {
+            rpcCallOpts = {
+                "cert": fs.readFileSync( joAccount.strPathSslCert, "utf8" ),
+                "key": fs.readFileSync( joAccount.strPathSslKey, "utf8" )
+            };
+        } else {
+            details.write(
+                cc.warning( "Will sign via SGX" ) + " " + cc.error( "without SSL options" ) +
+                "\n" );
+        }
+        const signerIndex = imaState.nNodeNumber;
+        await rpcCall.create( joAccount.strSgxURL, rpcCallOpts, async function( joCall, err ) {
+            if( err ) {
+                const strErrorMessage =
+                    strLogPrefix + cc.fatal( "CRITICAL ERROR:" ) +
+                    cc.error( " JSON RPC call to SGX failed, " +
+                        "RPC call was not created, error is: " ) +
+                    cc.warning( owaspUtils.extract_error_message( err ) ) +
+                    "\n";
+                log.write( strErrorMessage );
+                details.write( strErrorMessage );
+                if( joCall )
+                    await joCall.disconnect();
+                throw new Error(
+                    "JSON RPC call to SGX failed, RPC call was not created, error is: " +
+                    owaspUtils.extract_error_message( err )
+                );
+            }
+            const joCallSGX = {
+                "jsonrpc": "2.0",
+                "id": randomCallID(),
+                "method": "blsSignMessageHash",
+                "params": {
+                    "keyShareName": joAccount.strBlsKeyName,
+                    "messageHash": strMessageHash,
+                    "n": nParticipants,
+                    "t": nThreshold,
+                    "signerIndex": signerIndex + 0 // 1-based
```

Comment:
Could be replaced by `"signerIndex": signerIndex`

Reply:
No, `+0` here is trick for creating copy of value instead of variable reference))